bokehjs/src/coffee/document.coffee

```diff
@@ -614,6 +614,10 @@ class Document
         new_references[id] = value
 
     Document._initialize_references_json(references_json, old_references, new_references)
+    for id, model of new_references
+      model.attach_document(@)
+      if model.layoutable is true
+        @_add_layoutable(model)
```

Comment:
Before this change, you were doing `@_add_layoutable` only on the roots right, and with this change, you're doing it on all models anywhere in the tree?

The way we get to non-roots for attach_document / detach_document is that those methods are recursive (in has_props.coffee, we recurse into children).

However, aren't the methods called on the root model in `@_add_layoutable` already recursive, that is, `model.get_constraints` returns constraints for all children as well? At least box.coffee appears to work that way. So if you call `@_add_layoutable` on every child, I'd expect you're getting massive duplication of constraints?

Perhaps what's missing is a way for any model (not only one with children) to say "my constraints have changed, reload them" - this would need to propagate upward, so if any child of a container changes constraints, the container changes its aggregated constraints... the document would then listen to the root models, and reload their constraints when required.